### PR TITLE
[MRG+1] Fix compatibility with sklearn

### LINKED ISSUE #550: CV fit method should return self just like GridSearchCV in Sci-kit learn
In BayesSearchCV class, fit method does Not return self. It's inconsistent with the way sci-kit learn is built.
https://github.com/scikit-optimize/scikit-optimize/blob/b3cc7f6897c604988ab0a9e983adcd3cb11c301e/skopt/searchcv.py#L680
Here is fit method of GridSearchCV in sklearn. 
https://github.com/scikit-learn/scikit-learn/blob/8e599c68fb06967313d0e76a053062158cd90312/sklearn/model_selection/_search.py#L749

### PR DESCRIPTION
**Pull request by fcharras**
Doing *anything* in the `__init__` method of a sklearn estimator will most of the time break his compatibility with sklearn's clone, get_params and set_params, or the ability to fit it several times. It is in fact the case here, it would break if used in a Pipeline or other meta-objects in sklearn.

This PR is a quick (for now, dirty) proposal to fix the compatibility issues.

I've basically moved every lines of code that are not `self.foo = foo`, where foo is an arg, inside the fit() and add_space() methods.

Fixes #550 #551

### CODE REVIEW

**Comment by glouppe** [comment]
Thanks a lot! Let us know when it is ready for review.

**Comment by fcharras** [comment]
Haven't tested it yet I might try it within next week. Also, let's see if the checks passes. Aside from this the changes are minors so I guess it can be looked at. (mostly copy/paste code in init, to the top of add_space() and fit())

**Comment by betatim** [comment]
Thanks for spotting this and fixing it! To make bookkeeping easier could you add a line "Fixes #550 #551" in the top most comment for this PR? That way the issues get closed automatically when this gets merged. 📋

File: skopt/searchcv.py [hunk 1]
```diff
@@ -268,43 +268,17 @@ class BayesSearchCV(BaseSearchCV):
 
     """
 
-    def __init__(self, estimator, search_spaces=None, optimizer_kwargs=None,
+    def __init__(self, estimator, search_spaces=None, optimizer_kwrgs=None,
```
Comment (betatim) [inline skopt/searchcv.py]: `kwargs` is the standard in python so I think we should keep it

**Comment by betatim** [comment]
General comment: having to use `hasattr` this much feels wrong to my python brain. The scikit-learn design principle says we should store all arguments to `__init__` as attributes, and not compute anything until `fit()` is called. All computed attributes are stored as `self.blah_`. Do we need to have `add_spaces()` and `step()`? This adds to the API which is trying to copy the scikit-learn which we are trying to replicate. It is also the source of all the `hasattr`s?

Not sure what the conclusion is, but wanted to start a discussion.

File: skopt/searchcv.py (hunk 1 above)
Comment (fcharras) [reply skopt/searchcv.py]: Thank you for reviewing.

Agreed, it should be optimizer_kwargs, and all the current optimizer_kwargs should be optimizers_kwargs_.

I'll make a new, clean pass for variable names and refactoring. I think it's going to pass all checks this time.

**Comment by fcharras** [comment]
I very much agree with this, the current API that add_spaces and step offer could be dropped, it would make easier to stick to the standard sklearn API.

**Comment by betatim** [comment]
Let's see what the others think.

**Comment by iaroslav-ai** [comment]
First of all, thanks @fcharras for starting this :)
Yes, `add_spaces` and `step` should be removed - this will make things a lot easier. The idea was to have ask-and-tell interface for *SearchCV, but this can be achieved in a more simple / standard ways, like using callbacks #515.

**Comment by fcharras** [comment]
I think the most straightforward change would be to consider that `add_spaces` and `step` are privates, and only called during the `fit`. This would enable defining all the attributes `self.foo_ = ...` only at the beginning of the `fit`. Then the definitions in `step` and `add_spaces` and all the `hasattr` checks could be removed. Tests and documentation on `add_space` and `step` should be updated / deleted accordingly.

**Comment by iaroslav-ai** [comment]
@fcharras I agree - that would be the easiest solution. If you feel like / have time for implementing this I personally would be very happy if you do so :) Alternatively I will make a PR later.

**Comment by fcharras** [comment]
Thank you ! I'd like to keep pushing this PR, new commits soon.

**Comment by fcharras** [comment]
Here's a first shot. I finally totally deleted add_spaces because most of its content became obsolete with the change. Also deleted and moved some parts of step into fit.

File: examples/sklearn-gridsearchcv-replacement.ipynb [hunk 2]
```diff
@@ -2,10 +2,7 @@
  "cells": [
```
Comment (fcharras) [inline examples/sklearn-gridsearchcv-replacement.ipynb]: Woops, the changes in "metadata' are a mistake.

**Comment by fcharras** [comment]
I think it can be reviewed. All the tests and notebooks passes on my local machine. The notebooks are somewhat long to run on CircleCI though.

File: skopt/searchcv.py [hunk 3]
```diff
@@ -587,77 +509,37 @@ def _make_optimizer(self, params_space):
...
 
         return optimizer
 
-    def step(self, X, y, space_id, groups=None, n_jobs=1):
+    def _step(self, X, y, space_id, groups=None, n_jobs=1):
         """Generate n_jobs parameters and evaluate them in parallel.
 
         Having a separate function for a single step for search allows to
         save easily checkpoints for the parameter search and restore from
         possible failures.
```
Comment (betatim) [inline skopt/searchcv.py]: should we remove this comment (L521, 522, 523) as well?

File: skopt/searchcv.py [hunk 4]
```diff
@@ -673,35 +555,47 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
-                "calling fit method."
-            )
+        self.search_spaces_ = {}
+        for space, name in zip(search_spaces, list(range(len(search_spaces)))):
+            self.search_spaces_[name] = space
+
+        self.optimizer_kwargs_ = {} if self.optimizer_kwargs is None else\
+            self.optimizer_kwargs
+
+        # Instanciate optimizers for all the search spaces.
```
Comment (betatim) [inline skopt/searchcv.py]: Instanciate -> Instantiate

File: skopt/searchcv.py [hunk 5]
```diff
@@ -673,35 +555,47 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
-            raise ValueError(
-                "Please provide search space using `add_spaces` first before"
-                "calling fit method."
-            )
+        self.search_spaces_ = {}
+        for space, name in zip(search_spaces, list(range(len(search_spaces)))):
+            self.search_spaces_[name] = space
+
+        self.optimizer_kwargs_ = {} if self.optimizer_kwargs is None else\
+            self.optimizer_kwargs
```
Comment (betatim) [inline skopt/searchcv.py]: I'd make this into:
```python
if self.optimizer_kwargs is None:
  self.optimizer_kwargs_ = {}
else:
  self.optimizer_kwargs_ = self.optimizer_kwargs
```

instead of trying to fit it on one line, then having to use `\` to continue on a second line :)

File: skopt/searchcv.py [hunk 6]
```diff
@@ -673,35 +555,47 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
+        if isinstance(search_spaces, dict):
+            search_spaces = [search_spaces]
 
-        if len(self.search_spaces_) == 0:
-            raise ValueError(
-                "Please provide search space using `add_spaces` first before"
-                "calling fit method."
-            )
+        self.search_spaces_ = {}
+        for space, name in zip(search_spaces, list(range(len(search_spaces)))):
```
Comment (betatim) [inline skopt/searchcv.py]: Would `for space_id, space in enumerate(search_spaces):` not also work?

Why do we use a dictionary instead of a list if we use the position as key? We should add a comment that reminds us about the reason or change to a `list`?

File: skopt/searchcv.py [hunk 7]
```diff
@@ -673,35 +555,47 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
+            self.search_spaces_[name] = space
+
+        self.optimizer_kwargs_ = {} if self.optimizer_kwargs is None else\
+            self.optimizer_kwargs
+
+        # Instanciate optimizers for all the search spaces.
+        self.optimizer_ = {}
```
Comment (betatim) [inline skopt/searchcv.py]: -> `self.optimizers_` ?

File: skopt/searchcv.py [hunk 8]
```diff
@@ -722,10 +616,16 @@ def fit(self, X, y=None, groups=None):
             # do the optimization for particular search space
             while n_iter > 0:
                 # when n_iter < n_jobs points left for evaluation
-                n_jobs_adjusted = min(n_iter, self.n_jobs)
+                n_jobs_adjusted = min(n_iter, n_jobs)
```
Comment (betatim) [inline skopt/searchcv.py]: we need to add something that protects us from doing too many calls towards the end: when `n_iter > 0 and n_iter < n_jobs_adjusted`. Say we have `n_iter = 3` and `n_jobs = 4` we would evaluate one model too many currently.

File: skopt/searchcv.py [hunk 9]
```diff
@@ -722,10 +616,16 @@ def fit(self, X, y=None, groups=None):
...
 
-                self.step(
+                self._step(
                     X, y, space_id,
                     groups=groups, n_jobs=n_jobs_adjusted
                 )
                 n_iter -= n_jobs
+
+        # fit the best model if necessary
```
Comment (betatim) [inline skopt/searchcv.py]: -> "Refit the best model on the the whole dataset" ?

**Comment by betatim** [comment]
Should we add a test that checks that `BayesSearchCV` is compatible with the sklearn estimator semantics? So that this bug doesn't come back.

File: skopt/searchcv.py (hunk 3 above)
Comment (fcharras) [reply skopt/searchcv.py]: Ok, I think too that `_step` is mostly here for code clarity now. I don't know how things such as checkpoint would look, so removing.

File: skopt/searchcv.py (hunk 6 above)
Comment (fcharras) [reply skopt/searchcv.py]: The previous code was written with this dictionary, I don't know why. It seems right to use a list along with `enumerate`.

**Comment by fcharras** [comment]
Thank you for the review. I'm currently addressing it.

About the test for sklearn estimator semantics, I was wondering about using http://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html , I need some more time to investigate because I don't know exactly what this helper checks.

Another simpler option could be to add tests using clone / set_params / get_params.

In fact I haven't done those tests yet.

**Comment by fcharras** [comment]
I think this will address all the points you raised. Using lists to store search_space and optimizer make indeed the code a lot lighter. 

I provided a simple test for `clone()`, `set_params()` and `get_params()`, I'm not sure this is enough.

**Comment by fcharras** [comment]
As for `check_estimator` at https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html it appears that GridSearchCV doesn't pass the test. BayesSearchCV fail with the same error: `AssertionError: "0 feature\(s\) \(shape=\(3, 0\)\) while a minimum of \d* is required." does not match "n_splits=3 cannot be greater than the number of members in each class."`. I don't think we can use this.

**Comment by betatim** [comment]
A heads up on the CircleCI annoyance: #553 should fix that, once it is merged you can rebase your branch onto the latest master

**Comment by fcharras** [comment]
I tried to use it on a real world problem, it seems to work just fine with sklearn, however there's an issue with multi-processing: even when n_jobs = 1, a step tries to use all the cores. Is it that the underlying optimizers spawn as much processes as possible ?

**Comment by iaroslav-ai** [comment]
> even when n_jobs = 1, at some point it tries to use all the cores for a few seconds. Is it that the underlying optimizers spawn as much processes as possible ?

Yes, internally optimizer uses GaussianProcessRegressor by default. Inference with this estimator (used heavily during selection of next point to evaluate) runs on all cores if numpy by default runs on all cores (your's does that I assume). 

One can find [here](https://stackoverflow.com/questions/17053671/python-how-do-you-stop-numpy-from-multithreading) how to disable such behavior. Maybe would be good to turn this into an issue.

File: skopt/searchcv.py [hunk 10]
```diff
@@ -673,59 +550,75 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
-            n_features is the number of features.
+        X : array-like or sparse matrix, shape = [n_samples, n_features]
+            The training input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csc_matrix``.
 
         y : array-like, shape = [n_samples] or [n_samples, n_output]
-            Target relative to X for classification or regression;
+            Target relative to X for classification or regression (class
+            labels, as integers or strings).
```
Comment (iaroslav-ai) [inline skopt/searchcv.py]: Maybe write here "class labels should be integers or strings", otherwise someone might that any targets should be represented as integers or strings

File: skopt/searchcv.py [hunk 11]
```diff
@@ -673,59 +550,75 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
         # optimizer minimizes objective, hence provide negative score
         optimizer.tell(params, [-score for score in local_results])
 
-        # fit the best model if necessary
-        if self.refit:
```
Comment (iaroslav-ai) [inline skopt/searchcv.py]: :+1:

File: skopt/tests/test_searchcv.py [hunk 12]
```diff
@@ -42,9 +41,6 @@ def test_searchcv_runs(surrogate, n_jobs):
         X, y, train_size=0.75, random_state=0
     )
 
-    # None search space is only supported when only `step` function is used
-    assert_raises(ValueError, BayesSearchCV(SVC(), None).fit, (X, y))
```
Comment (iaroslav-ai) [inline skopt/tests/test_searchcv.py]: Maybe instead of this add test with zero length list?

File: skopt/searchcv.py [hunk 13]
```diff
@@ -673,59 +550,75 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
         """Run fit on the estimator with randomly drawn parameters.
 
         Parameters
         ----------
-        X : array-like, shape = [n_samples, n_features]
-            Training vector, where n_samples in the number of samples and
-            n_features is the number of features.
+        X : array-like or sparse matrix, shape = [n_samples, n_features]
+            The training input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
```
Comment (iaroslav-ai) [inline skopt/searchcv.py]: Conversion to np.float32 is usually done with a particular estimator that is used with BayesSearchCV, but not BayesSearchCV itself. So one can have something like 
```python
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import Lasso

from skopt import BayesSearchCV

# strings as input
X = ['a'*np.random.randint(1,10) for i in range(100)]
# len of string should be estimated output
y = [len(x) for x in X]

# transformer on strings that converts them to features
class MyTransformer(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        X = [[len(x)] for x in X]
        X = np.array(X)
        return X

estimator = make_pipeline(
    MyTransformer(),
    Lasso()
)

model = BayesSearchCV(
    estimator=estimator,
    search_spaces={
        'lasso__alpha': (0.1, 10.0)
    },
    n_iter=8
)

model.fit(X, y)
print(model.score(X, y))
```
where inputs are strings. So consider deleting the comment about conversion of data.

**Comment by fcharras** [comment]
Thanks for the review.

> runs on all cores if numpy by default runs on all cores (your's does that I assume).

I'm currently debugging it. It's indeed the `optimizer.tell` step that spawn that much threads. I'm pretty sure I have a standard python install / standard default ubuntu libraries for blas, mpi and such, I thought those wouldn't be parallelized by defaults ?

**Comment by iaroslav-ai** [comment]
My reviews are nitpicks for most part :)
It is great that you removed `step` and other related stuff (I mean now code is ~100 lines shorter, just that is awesome). Originally the idea was to have ask-and-tell sorta interface for BayesSearchCV, but after all the feedback I do not find the idea so good (maybe I am just too much into ask - and - tell interface :D )

**Comment by fcharras** [comment]
I've adressed your comments. 

The purpose of the PR is to fix the compatibility with sklearn and I think it is fulfilled now (I've tested it more extensively during the week with complex sklearn objects like nested Pipelines, calibrators,... no issue encountered). There might be other details that can be improved that remain from the last version (e.g documentation) that I missed. 

I'd like to continue polishing the class in new PRs. Some might demand more discussion or iterations (I'd like to reconsider the inheritance to BaseSearchCV, or at least how the _fit is defined, propose a change to the API and documentation so that `n_iter` becomes `n_iter / n_jobs`, set `pre_dispatch` to `n_jobs` and remove it from the `__init__` args, and consider using a single Parallel instance,...).

> After all the feedback I do not find the idea of ask and tell for *SearchCV so good, maybe I was just too much into it :D

As a heavy sklearn user, I was expecting that BayesSearchCV would _really_ be a drop-in replacement for GridSearchCV in all my pipelines. I think others will have the same expectations, and if the performance is good, this tool could be very popular once the PR is merged and the compatibility is flawless.

But the sklearn API and the compatibility requirements are somewhat inflexible and exclusive, so that comes at the cost of other possibly creative APIs like the ask-and-tell interface. In my opinion this is good too but it should be defined in a separate class.

So I think it's not about _not wanting_ ask-and-tell, but more about _really wanting_ sklearn compatibility, which is in my opinion one of the sanest framework.

**Comment by glouppe** [comment]
Thanks a lot for your efforts! I am marking this for 0.5, as this is super important.

**Comment by betatim** [comment]
Woah! I think I agree with most of what @fcharras wrote in https://github.com/scikit-optimize/scikit-optimize/pull/551#issuecomment-343646538

File: skopt/searchcv.py [hunk 14]
```diff
@@ -44,7 +44,7 @@ class BayesSearchCV(BaseSearchCV):
         or ``scoring`` must be passed.
 
     search_spaces : dict, list of dict or list of tuple containing
-        (dict, int), or None.
+        (dict, int).
         One of 4 following cases:
```
Comment (betatim) [inline skopt/searchcv.py]: -> "One of these cases:" (don't have four anymore and this way we won't have to update the text again when a new one is added ;))

File: skopt/searchcv.py [hunk 15]
```diff
@@ -319,6 +291,11 @@ def _check_search_space(self, search_space):
         if isinstance(search_space, dict):
             search_space = [search_space]
 
+        if len(self.search_spaces) == 0:
```
Comment (betatim) [inline skopt/searchcv.py]: In the rest of the method we use the `search_space` passed as argument. We should be consistent

File: skopt/searchcv.py [hunk 16]
```diff
@@ -268,43 +266,17 @@ class BayesSearchCV(BaseSearchCV):
...
-        # instances for different search spaces. `str(space)` is used as key
-        # as `space` is a dict, which is unhashable. however, str(space)
-        # is fixed and unique if space is not changed.
-        self.optimizer_ = {}
-        self.cv_results_ = defaultdict(list)
-
-        self.best_index_ = None
-        self.multimetric_ = False
+        self.optimizer_kwargs = optimizer_kwargs
+        self._check_search_space(self.search_spaces)
```
Comment (betatim) [inline skopt/searchcv.py]: What do you think of making this `self.search_spaces = self._check_search_space(search_spaces)` and removing L275?

In real scikit-learn style we should not check the arguments here but instead wait till `fit` is called :-/ Maybe that is actually the best thing to do. http://scikit-learn.org/stable/developers/contributing.html#instantiation WDYT?

File: skopt/searchcv.py (hunk 15 above)
Comment (betatim) [reply skopt/searchcv.py]: When does this happen? A user passes an empty list/dict?

File: skopt/searchcv.py [hunk 17]
```diff
@@ -319,6 +291,11 @@ def _check_search_space(self, search_space):
         if isinstance(search_space, dict):
             search_space = [search_space]
 
+        if len(self.search_spaces) == 0:
+            raise ValueError(
+                "Please provide at least one search space"
```
Comment (betatim) [inline skopt/searchcv.py]: -> `"The search_spaces parameter has to contain at least one search space not %s" % search_space` I find it helps the user (aka me debugging at 3am) a lot to have the parameter name and the current value of it in the message.

File: skopt/searchcv.py [hunk 18]
```diff
@@ -673,59 +550,73 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
 
-        for space_id in sorted(self.search_spaces_.keys()):
-            elem = self.search_spaces_[space_id]
-
+        for search_space, optimizer in zip(search_spaces, optimizers):
             # if not provided with search subspace, n_iter is taken as
             # self.n_iter
-            if isinstance(elem, tuple):
-                space, n_iter = elem
+
```
Comment (betatim) [inline skopt/searchcv.py]: no blank line here please

File: skopt/searchcv.py [hunk 19]
```diff
@@ -673,59 +550,73 @@ def step(self, X, y, space_id, groups=None, n_jobs=1):
...
-        if len(self.search_spaces_) == 0:
-            raise ValueError(
-                "Please provide search space using `add_spaces` first before"
-                "calling fit method."
-            )
+        if self.optimizer_kwargs is None:
+            self.optimizer_kwargs_ = {}
+        else:
+            self.optimizer_kwargs_ = self.optimizer_kwargs
+
```
Comment (betatim) [inline skopt/searchcv.py]: we need to call [`check_random_state`](http://scikit-learn.org/stable/developers/contributing.html#random-numbers) here and get random ints from it (`rng.randint(0, np.iinfo(np.int32).max)`) for each of our `Optimizer` instances. Otherwise the results won't be reproducible.

Comment (betatim) [reply skopt/searchcv.py]: We should add a test doing a short optimisation with two different instances of `BayesSearchCV` to check things are indeed reproducible.

File: skopt/searchcv.py (hunk 16 above)
Comment (fcharras) [reply skopt/searchcv.py]: I guess it's OK to run code in the `__init__` that does not alter attributes, a similar check is performed in `GridSearchCV` https://github.com/scikit-learn/scikit-learn/blob/f3320a6f1ab2d060fd49c94a6e3386291a684ec7/sklearn/model_selection/_search.py#L1077

File: skopt/searchcv.py (hunk 15 above)
Comment (fcharras) [reply skopt/searchcv.py]: I'm changing it a little so that uses search_space, good catch. It will also be triggered both with an input empty dict or empty list. However it wouldn't trigger with a list of empty dicts. It looks like check_dimension (L333) would raise an error for any empty dict anyway.

This method  `check_search_space` already looks very similar to GridSearchCV's check_param_grid https://github.com/scikit-learn/scikit-learn/blob/f3320a6f1ab2d060fd49c94a6e3386291a684ec7/sklearn/model_selection/_search.py#L343 .

File: skopt/searchcv.py (hunk 17 above)
Comment (fcharras) [reply skopt/searchcv.py]: thank you for providing the code 😸

**Comment by fcharras** [comment]
Thank you for the review, those commits should address it

**Comment by fcharras** [comment]
I've been thinking about the test on reproducibility, using a SVM as a base estimator might not make the safest test because of its stability regardless of the parameters. I've changed it to a random forest, but now the test passes... only half the time.

**Comment by fcharras** [comment]
Added a better test for the reproducibility. The test is definitely not passing, results are not reproducible... isn't https://github.com/scikit-optimize/scikit-optimize/pull/551/files#diff-5be74bdba5232db1bd3137d5827653f6R583 correct though ?

**Comment by betatim** [comment]
To me `test_searchcv_reproducibility` looks like it should work and test what we want it to test. So if it reliability passes that is good ;)

File: skopt/tests/test_searchcv.py [hunk 20]
```diff
@@ -124,3 +124,102 @@ def test_searchcv_runs_multiple_subspaces():
...
+    )
+
+    opt.fit(X_train, y_train)
+    best_estimator = opt.best_estimator_
+
+    opt2 = clone(opt).fit(X_train, y_train)
+    best_estimator2 = opt2.best_estimator_
+
+    for attr in ('C', 'gamma', 'degree', 'kernel'):
+        assert_equal(getattr(best_estimator, attr),
```
Comment (betatim) [inline skopt/tests/test_searchcv.py]: with pytest you can do `assert getattr(best_estimator, attr) == getattr(best_estimator2, attr)`. I would prefer explicitly writing out the four lines instead of having a for loop. In the interest of keeping our tests as simple as possible in terms of complexity and it makes it a bit simpler to see in the output which parameter is different.

**Comment by betatim** [comment]
I think this looks pretty good. Maybe change the for loop in the test to satisfy my nitpicking.

File: skopt/tests/test_searchcv.py [hunk 21]
```diff
@@ -124,3 +124,102 @@ def test_searchcv_runs_multiple_subspaces():
...
+            'degree': Integer(1, 8),
+            'kernel': Categorical(['linear', 'poly', 'rbf']),
+        },
+        n_iter=11, random_state=random_state
+    )
+
+    opt.fit(X_train, y_train)
+    best_estimator = opt.best_estimator_
+
+    opt2 = clone(opt).fit(X_train, y_train)
```
Comment (iaroslav-ai) [inline skopt/tests/test_searchcv.py]: I wonder if there might be a situation where the test will pass because the estimator being cloned was already fitted? That is, when the estimator is cloned before calling `fit` method, the test would fail. From looking at definition of clone method, I am not sure. If you think this might happen, consider adding another copy before fitting, and comparing it with the rest.

File: skopt/searchcv.py [hunk 22]
```diff
@@ -268,43 +268,17 @@ class BayesSearchCV(BaseSearchCV):
 
     """
 
-    def __init__(self, estimator, search_spaces=None, optimizer_kwargs=None,
+    def __init__(self, estimator, search_spaces, optimizer_kwargs=None,
```
Comment (iaroslav-ai) [inline skopt/searchcv.py]: Something that is probably out of scope for this PR: I realize now is that `optimizer_kwargs` are the same for all subspaces. However, one might want these to be different for different subspaces. An example would be that I want optimizer to try a few initial points, that I suspect already to be good for some subspace. This could be done by specifying kwargs for optimizer; However, in case of multiple subspaces this is not possible as kwargs for those are the same. Similarly, one might want to use decision trees based optimizers for subspaces with large number of iterations and models with fast fitting. 
So kwargs could be specified per subspace, like iterations number. For now one can simply have a separate BayesSearchCv per subspace. 
Edit: maybe lets leave this out, in order to avoid everlasting PR's :)

**Comment by glouppe** [comment]
I had several quick looks over the past two weeeks. This now seems like a significant improvement over master. All the little things can be addressed later in smaller PR in my opinion.

+0.5 for merge on my side (since I have not reviewed the PR in full details)

**Comment by iaroslav-ai** [comment]
Yes, I am fine also with merging this as it is.

**Comment by fcharras** [comment]
Thank you! I think if the test on reproducibility passes here, then it passes consistently and there were some issue with my setup on my side that made me think it does not pass.

I can check again by tonight and maybe you could try it too to be sure. About the other little improvements (i have other in mind too) I could write some documentation that list them that could be addressed in further PR.

File: skopt/tests/test_searchcv.py (hunk 21 above)
Comment (fcharras) [reply skopt/tests/test_searchcv.py]: I don't think this is an issue, I think clone always return a new un-fitted copy of the estimator, regardless of wether the input estimator is fitted or not. I'll double check it.

Comment (fcharras) [reply skopt/tests/test_searchcv.py]: Indeed, you can check https://github.com/scikit-learn/scikit-learn/blob/19eb62ff/sklearn/base.py#L30 and https://github.com/scikit-learn/scikit-learn/blob/19eb62ff/sklearn/base.py#L198 : `clone` return a fresh instance of the same class, and get the parameters from the signature of the `__init__` of the cloned instance.

File: skopt/tests/test_searchcv.py (hunk 20 above)
Comment (fcharras) [reply skopt/tests/test_searchcv.py]: I changed to assert, but using `assert_equal` seems nicer because it also print the values that should have been equal (ex: `43.5 = 1003.32`)

File: skopt/searchcv.py (hunk 22 above)
Comment (fcharras) [reply skopt/searchcv.py]: I think this could be done in further PR too, I'd like to propose some documentation with further possible improvements.

**Comment by fcharras** [comment]
I checked again, the last test is definitely not passing on my side, but definitely passing on the integrated tests here... On my side I ran it inside an ipython terminal in spyder. It's failing 100% of the time.

File: skopt/searchcv.py (hunk 22 above)
Comment (betatim) [reply skopt/searchcv.py]: @iaroslav-ai can you make an issue for this so we don't lose the idea?

**Comment by betatim** [comment]
Tests pass on travis and on my laptop -> merging!

Thanks a lot for these improvements, as well as your patience for my never ending comments :)

**Comment by fcharras** [comment]
Thank you!!

**Comment by betatim** [comment]
@fcharras could you send me an email (<mygithubusername>@google.com)? We'd like to get in touch with you and our internet stalking skills aren't good enough to find your email address :)
